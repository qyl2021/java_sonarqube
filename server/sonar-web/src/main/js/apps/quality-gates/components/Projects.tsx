/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { find, without } from 'lodash';
import * as React from 'react';
import {
  associateGateWithProject,
  dissociateGateWithProject,
  searchProjects
} from '../../../api/quality-gates';
import SelectList, { Filter } from '../../../components/SelectList/SelectList';
import { translate } from '../../../helpers/l10n';

interface Props {
  canEdit?: boolean;
  organization?: string;
  qualityGate: T.QualityGate;
}

export interface SearchParams {
  gateId: number;
  organization?: string;
  page: number;
  pageSize: number;
  query?: string;
  selected: string;
}

interface State {
  lastSearchParams: SearchParams;
  listHasBeenTouched: boolean;
  projects: Array<{ id: string; name: string; selected: boolean }>;
  projectsTotalCount?: number;
  selectedProjects: string[];
}

const PAGE_SIZE = 100;

export default class Projects extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);

    this.state = {
      lastSearchParams: {
        gateId: props.qualityGate.id,
        organization: props.organization,
        page: 1,
        pageSize: PAGE_SIZE,
        query: '',
        selected: Filter.Selected
      },
      listHasBeenTouched: false,
      projects: [],
      selectedProjects: []
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.fetchProjects(this.state.lastSearchParams);
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchProjects = (searchParams: SearchParams, more?: boolean) =>
    searchProjects({
      ...searchParams,
      query: searchParams.query !== '' ? searchParams.query : undefined
    }).then(data => {
      if (this.mounted) {
        this.setState(prevState => {
          const projects = more ? [...prevState.projects, ...data.results] : data.results;
          const newSelectedProjects = data.results
            .filter(project => project.selected)
            .map(project => project.id);
          const selectedProjects = more
            ? [...prevState.selectedProjects, ...newSelectedProjects]
            : newSelectedProjects;

          return {
            lastSearchParams: searchParams,
            listHasBeenTouched: false,
            projects,
            projectsTotalCount: data.paging.total,
            selectedProjects
          };
        });
      }
    });

  handleLoadMore = () =>
    this.fetchProjects(
      {
        ...this.state.lastSearchParams,
        page: this.state.lastSearchParams.page + 1
      },
      true
    );

  handleReload = () =>
    this.fetchProjects({
      ...this.state.lastSearchParams,
      page: 1
    });

  handleSearch = (query: string, selected: string) =>
    this.fetchProjects({
      ...this.state.lastSearchParams,
      page: 1,
      query,
      selected
    });

  handleSelect = (id: string) =>
    associateGateWithProject({
      gateId: this.props.qualityGate.id,
      organization: this.props.organization,
      projectId: id
    }).then(() => {
      if (this.mounted) {
        this.setState(state => ({
          listHasBeenTouched: true,
          selectedProjects: [...state.selectedProjects, id]
        }));
      }
    });

  handleUnselect = (id: string) =>
    dissociateGateWithProject({
      gateId: this.props.qualityGate.id,
      organization: this.props.organization,
      projectId: id
    }).then(() => {
      if (this.mounted) {
        this.setState(state => ({
          listHasBeenTouched: true,
          selectedProjects: without(state.selectedProjects, id)
        }));
      }
    });

  renderElement = (id: string): React.ReactNode => {
    const project = find(this.state.projects, { id });
    return project === undefined ? id : project.name;
  };

  render() {
    return (
      <SelectList
        elements={this.state.projects.map(project => project.id)}
        elementsTotalCount={this.state.projectsTotalCount}
        labelAll={translate('quality_gates.projects.all')}
        labelSelected={translate('quality_gates.projects.with')}
        labelUnselected={translate('quality_gates.projects.without')}
        needReload={
          this.state.listHasBeenTouched && this.state.lastSearchParams.selected !== Filter.All
        }
        onLoadMore={this.handleLoadMore}
        onReload={this.handleReload}
        onSearch={this.handleSearch}
        onSelect={this.handleSelect}
        onUnselect={this.handleUnselect}
        readOnly={!this.props.canEdit}
        renderElement={this.renderElement}
        selectedElements={this.state.selectedProjects}
      />
    );
  }
}
