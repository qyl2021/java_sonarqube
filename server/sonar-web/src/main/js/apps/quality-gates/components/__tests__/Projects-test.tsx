/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import Projects, { SearchParams } from '../Projects';
import SelectList, { Filter } from '../../../../components/SelectList/SelectList';
import { mockQualityGate } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import {
  searchProjects,
  associateGateWithProject,
  dissociateGateWithProject
} from '../../../../api/quality-gates';

jest.mock('../../../../api/quality-gates', () => ({
  searchProjects: jest.fn().mockResolvedValue({
    paging: { pageIndex: 1, pageSize: 3, total: 55 },
    results: [
      { id: 'test1', name: 'test1', selected: false },
      { id: 'test2', name: 'test2', selected: false },
      { id: 'test3', name: 'test3', selected: true }
    ]
  }),
  associateGateWithProject: jest.fn().mockResolvedValue({}),
  dissociateGateWithProject: jest.fn().mockResolvedValue({})
}));

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot();
  expect(searchProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      page: 1
    })
  );

  wrapper.setState({ listHasBeenTouched: true });
  expect(wrapper.find(SelectList).props().needReload).toBe(true);

  wrapper.setState({ lastSearchParams: { selected: Filter.All } as SearchParams });
  expect(wrapper.find(SelectList).props().needReload).toBe(false);
});

it('should handle reload properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleReload();
  expect(searchProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      page: 1
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(false);
});

it('should handle search reload properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleSearch('foo', Filter.Selected);
  expect(searchProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      page: 1,
      query: 'foo',
      selected: Filter.Selected
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(false);
});

it('should handle load more properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleLoadMore();
  expect(searchProjects).toHaveBeenCalledWith(
    expect.objectContaining({
      page: 2
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(false);
});

it('should handle selection properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleSelect('toto');
  await waitAndUpdate(wrapper);
  expect(associateGateWithProject).toHaveBeenCalledWith(
    expect.objectContaining({
      projectId: 'toto'
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(true);
});

it('should handle deselection properly', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.instance().handleUnselect('tata');
  await waitAndUpdate(wrapper);
  expect(dissociateGateWithProject).toHaveBeenCalledWith(
    expect.objectContaining({
      projectId: 'tata'
    })
  );
  expect(wrapper.state().listHasBeenTouched).toBe(true);
});

function shallowRender(props: Partial<Projects['props']> = {}) {
  return shallow<Projects>(<Projects qualityGate={mockQualityGate()} {...props} />);
}
