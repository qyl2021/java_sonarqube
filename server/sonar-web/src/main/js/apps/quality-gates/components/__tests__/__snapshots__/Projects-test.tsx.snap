// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<SelectList
  elements={
    Array [
      "test1",
      "test2",
      "test3",
    ]
  }
  elementsTotalCount={55}
  labelAll="quality_gates.projects.all"
  labelSelected="quality_gates.projects.with"
  labelUnselected="quality_gates.projects.without"
  needReload={false}
  onLoadMore={[Function]}
  onReload={[Function]}
  onSearch={[Function]}
  onSelect={[Function]}
  onUnselect={[Function]}
  readOnly={true}
  renderElement={[Function]}
  selectedElements={
    Array [
      "test3",
    ]
  }
/>
`;
